Index: /vendor/magento/module-customer/Controller/Account/LoginPost.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- /vendor/magento/module-customer/Controller/Account/LoginPost.php	(revision 27f2363d53c441fc4377d61578a3651c0a504f3e)
+++ /vendor/magento/module-customer/Controller/Account/LoginPost.php	(revision 7baa634062cbee1d64503dc23e0bef7676bd8617)
@@ -67,6 +67,11 @@
      */
     private $cookieMetadataManager;

+    /**
+     * @var CustomerUrl
+     */
+    private $customerUrl;
+
     /**
      * @param Context $context
      * @param Session $customerSession
@@ -200,10 +205,11 @@
                     }
                 } catch (EmailNotConfirmedException $e) {
                     $value = $this->customerUrl->getEmailConfirmationUrl($login['username']);
-                    $message = __(
+                    $this->messageManager->addError(__(
                         'This account is not confirmed. <a href="%1">Click here</a> to resend confirmation email.',
                         $value
-                    );
+                    ));
+                    $this->session->setUsername($login['username']);
                 } catch (AuthenticationException $e) {
                     $message = __(
                         'The account sign-in was incorrect or your account is disabled temporarily. '
Index: /dev/tests/integration/testsuite/Magento/Customer/Controller/Account/LoginPostTest.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- /dev/tests/integration/testsuite/Magento/Customer/Controller/Account/LoginPostTest.php	(revision 27f2363d53c441fc4377d61578a3651c0a504f3e)
+++ /dev/tests/integration/testsuite/Magento/Customer/Controller/Account/LoginPostTest.php	(revision 7baa634062cbee1d64503dc23e0bef7676bd8617)
@@ -13,6 +13,7 @@
 use Magento\Framework\App\Request\Http as HttpRequest;
 use Magento\Framework\Message\MessageInterface;
 use Magento\Framework\Url\EncoderInterface;
+use Magento\Framework\UrlInterface;
 use Magento\TestFramework\TestCase\AbstractController;

 /**
@@ -98,6 +99,8 @@
     }

     /**
+     * Tests correct message appears when Sign In with unconfirmed Customer.
+     *
      * @magentoDataFixture Magento/Customer/_files/customer_confirmation_config_enable.php
      * @magentoDataFixture Magento/Customer/_files/unconfirmed_customer.php
      *
@@ -105,17 +108,25 @@
      *
      * @return void
      */
-    public function testLoginWithUnconfirmedPassword(): void
+    public function testLoginWithUnconfirmedCustomer(): void
     {
-        $this->markTestSkipped('Blocked by MC-31370.');
         $email = 'unconfirmedcustomer@example.com';
+        $urlBuilder = $this->_objectManager->get(UrlInterface::class);
         $this->prepareRequest($email, 'Qwert12345');
         $this->dispatch('customer/account/loginPost');
         $this->assertEquals($email, $this->session->getUsername());
-        $this->assertSessionMessages(
-            $this->equalTo([(string)__('This account is not confirmed. Click here to resend confirmation email.')]),
-            MessageInterface::TYPE_ERROR
+        $expectedMessage = __(
+            'This account is not confirmed. <a href="'
+            . $urlBuilder->getUrl('customer/account/confirmation', ['_query' => ['email' => $email]])
+            . '">Click here</a> to resend confirmation email.'
         );
+        $this->assertSessionMessages($this->equalTo([(string)$expectedMessage]), MessageInterface::TYPE_ERROR);
+        $errorMessage = current($this->getMessages(MessageInterface::TYPE_ERROR));
+        $entities = ['&lt;', '&gt;', '&quot;', '&#039;', '&amp;'];
+        foreach ($entities as $entity) {
+            $this->assertNotContains($entity, $errorMessage);
+        }
+        $this->assertRedirect($this->stringContains('customer/account/login'));
     }

     /**
Index: /vendor/magento/module-customer/Test/Unit/Controller/Account/LoginPostTest.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- /vendor/magento/module-customer/Test/Unit/Controller/Account/LoginPostTest.php	(revision 7baa634062cbee1d64503dc23e0bef7676bd8617)
+++ /vendor/magento/module-customer/Test/Unit/Controller/Account/LoginPostTest.php	(revision b44f9ba21ca0cba017277c715fbc715a256fdc3b)
@@ -557,7 +557,7 @@
                     $url
                 );
                 $this->messageManager->expects($this->once())
-                    ->method('addErrorMessage')
+                    ->method('addError')
                     ->with($message)
                     ->willReturnSelf();

